src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -2609,7 +2609,7 @@ private BoundStatement BindReturn(ReturnStatementSyntax syntax, DiagnosticBag di
 
             if (hasErrors)
             {
-                return new BoundReturnStatement(syntax, refKind, arg, hasErrors: true);
+                return new BoundReturnStatement(syntax, sigRefKind, arg, hasErrors: true);
```

Comment:
>sigRefKind [](start = 56, length = 10)

Are all possible consumers ready to deal with ```RefReadOnly``` value here. For example, ```VisitReturnStatementNoAdjust``` and the methods it calls in AbstractFlowPass.

Reply:
Expression-bodied members already produce `RefReadOnly` today. Are there any code paths unique to block-bodied members that need to be reviewed separately?

Reply:
> Expression-bodied members already produce RefReadOnly today. Are there any code paths unique to block-bodied members that need to be reviewed separately?

That was likely an oversight and probably was not tested properly. I do not feel comfortable simply expanding that to a more common case. I think that all consumers should be located, targeted tests for them should be added, appropriate adjustments to implementation should be made. For example, In AbstractFlowPass:
```
        protected virtual BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement node)
        {
            VisitRvalue(node.ExpressionOpt);

            // byref return is also a potential write
            if (node.RefKind != RefKind.None)
            {
                WriteArgument(node.ExpressionOpt, node.RefKind, method: null);
            }

            return null;
        }
```

It looks like in DefiniteAssignmentPass implementation of WriteArgument handles ```RefKind.Ref```, but doesn't handle ```RefKind.RefReadOnly```.

```
        protected override void WriteArgument(BoundExpression arg, RefKind refKind, MethodSymbol method)
        {
            if (refKind == RefKind.Ref)
            {
                // Though the method might write the argument, in the case of ref arguments it might not,
                // thus leaving the old value in the variable.  We model this as a read of the argument
                // by the method after the invocation.
                CheckAssigned(arg, arg.Syntax);
            }

            Assign(arg, value: null);

            // Imitate Dev10 behavior: if the argument is passed by ref/out to an external method, then
            // we assume that external method may write and/or read all of its fields (recursively).
            // Strangely, the native compiler requires the "ref", even for reference types, to exhibit
            // this behavior.
            if (refKind != RefKind.None && ((object)method == null || method.IsExtern))
            {
                MarkFieldsUsed(arg.Type);
            }
        }
```

I think all consumers like that should be examined and tested. I guess, by default we should expect ``RefKind.Ref``` and ```RefKind.RefReadOnly``` produce the same result, unless there is a reason for it to be different.
  

---
In reply to: [292253519](https://github.com/dotnet/roslyn/pull/36271#discussion_r292253519) [](ancestors = 292253519)

Reply:
To minimize the cost of dealing with/testing consumers, we could do the following:
- Add TargetRefKind property to the BoundReturnStatement and store information from the signature there instead
- Adjust implementation around RefKind (align it with return statement) for expression-bodied members, assuming no existing tests fail.
- IOperation implementation uses one value or the other depending on whether there is a mismatch.
- TargetRefKind remains unused with the compiler outside of IOperation factory.

---
In reply to: [292576190](https://github.com/dotnet/roslyn/pull/36271#discussion_r292576190) [](ancestors = 292576190,292253519)

Reply:
We can even skip this part: "Adjust implementation around RefKind (align it with return statement) for expression-bodied members". And instead open an issue for the compiler devs to investigate and decide what would be the appropriate way to deal with it.

---
In reply to: [292600139](https://github.com/dotnet/roslyn/pull/36271#discussion_r292600139) [](ancestors = 292600139,292576190,292253519)